modules/foreign-state-oracle/src/lib.rs

```diff
@@ -0,0 +1,277 @@
+// This file is part of Acala.
+
+// Copyright (C) 2020-2022 Acala Foundation.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Foreign State Oracle.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{
+	dispatch::{Dispatchable, GetDispatchInfo, PostDispatchInfo},
+	pallet_prelude::*,
+	traits::{Currency, ExistenceRequirement, NamedReservableCurrency},
+	transactional, PalletId,
+};
+use frame_system::pallet_prelude::*;
+use module_support::ForeignChainStateQuery;
+use primitives::{Balance, ReserveIdentifier};
+use sp_runtime::{
+	traits::{AccountIdConversion, BlockNumberProvider, One, Saturating, Scale},
+	ArithmeticError,
+};
+use sp_std::{prelude::Vec, vec};
+
+mod mock;
+mod tests;
+
+pub use module::*;
+
+// Unique Identifier for each query
+pub type QueryIndex = u64;
+
+// Origin with arbitrary bytes included
+#[derive(PartialEq, Eq, Clone, RuntimeDebug, Encode, Decode, TypeInfo)]
+pub struct RawOrigin {
+	data: Vec<u8>,
+}
+
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub struct VerifiableCall<Call, BlockNumber, Balance> {
```

Comment:
Not sure if `VerifiableCall` is the right name. Where is the Verifiable coming from? I think it is more of a `RelayRequest` to request to relay some foreign state

Reply:
Renamed to RelayQueryRequest